contributors/design-proposals/storage/volume-reconstruct.md

```diff
@@ -0,0 +1,88 @@
+	ReadOnly         bool
+}
+```
+Both Volume and PersistentVolume has a name and volume source. Different volume plugins have different representation of their sources. 
+Volume spec is constructed when adding pod volume information into desired state.
+The main use of this spec is to get the volume source name to construct unique volume name (which typicall includes plugin name and volume source name). 
+The unique volume name is used as a key in actual/desired state. The volume source name is also used on construct global mount path (device mount path)
+for volume mounter. When volume is teared down, unmounter uses the same function to construct the mount path from the spec again.
+Different from global mount path, the bind mount path uses plugin name and spec name (not the volume source name). 
+
+So look more closely, you can see for only the purpose of volume tear down, if we have a record of those mount paths (bind and global) in actual state, there is no need to 
+construct those paths from the volume spec again. In case of kubelet restarts, those mount paths could be scanned from the mount lists left on the disk. (bind mount path has
+a path structure which can be easily idenetified and global mount path could be obtained from the mount reference of the bind mount path). This observation leads to our first
+possible solution.
+
+### Remove the dependency of volume spec. 
+Modify the data structure in actual state to keep a record of the mount paths so that volume tear down process (unmount) can use those records instead of reconstructing them
+directly from volume spec. In case the kubelet restarts, reconstruction process just need to find out the left mounting paths and mark them into the actual state. Even if the
+volume spec cannot be reconstructed correctly, volume tear down process can still work. 
+
+But this solution still has an issue. Because some plugins cannot reconstruct volume spec correctly, which means the generated unique volume name might not match with the real one.
+In case another new pod is added to use the same volume, two volume operations on the same device might happen concurrently and cause race condition because they have different
+unique volume names.
+
+### Cleanup mounts directly without reconstruction.
+This approach will try to clean up mounts that are no longer needed before reconciler starts to work. It scanns the pod volume directories when kubelet restarts to find the mount
+paths (bind and global). If the pod volume is no longer exist in desired state, clean up the mounts. To avoid long delay or mount hanging problem, we can use operation executor to
+start a new routine. 
+
+The potentical issue of this approach is that cleanup process might fail and
+reconciler will not try to do this process again.
+
+## Proposal
+
+The current proposal is trying to take the advantages from the two possible
+solutions mentioned above. 
+1. If volume plugins can support reconstruction well (i.e., volume spec can
```

Comment:
Can you add some detail about how this will be determined? It would be nice to see a little more detail in the proposal about how this is planned to be implemented.

Reply:
Currently, for volumes such as AWS, GCE, the volume source name can be retrieved by checking the mount reference.